packages/firestore/src/local/local_store.ts

```diff
@@ -918,7 +936,39 @@ export class LocalStore {
     return promiseChain;
   }
 
-  getActiveClients(): Promise<ClientId[]> {
-    return this.persistence.getActiveClients();
+  /**
+   *  Stores the mutation keys associated with the given batch ID. The mutation
+   *  keys can later be retrieved via `lookupMutationDocuments`.
+   */
+  // PORTING NOTE: Multi-tab only.
+  private insertMutationKeys(batchId: BatchId, keys: DocumentKeySet): void {
+    this.localMutationBatchKeys[batchId] = keys;
+  }
+
+  /**
+   * Looks up mutation keys for the given batch ID. If the mutation is not
+   * known locally (e.g. when another client added it), this method looks up the
+   * mutation from persistence.
+   */
+  // PORTING NOTE: Multi-tab only.
+  private lookupAndStoreMutationKeys(
+    txn: PersistenceTransaction,
+    batchId: BatchId
+  ): PersistencePromise<DocumentKeySet | null> {
+    if (this.localMutationBatchKeys[batchId]) {
+      return PersistencePromise.resolve(this.localMutationBatchKeys[batchId]);
+    } else {
+      return this.mutationQueue
+        .lookupMutationBatch(txn, batchId)
+        .next(batch => {
+          if (batch) {
+            const keys = batch.keys();
+            this.insertMutationKeys(batchId, keys);
```

Comment:
FWIW, I think this code would be more readable if you got rid of the insertMutationKeys() method. That way it's clear that you're reading (on line 958) and writing to (here) the same data (`this.localMutationBatchKeys`)

Reply:
No longer applicable. I mostly did this so it was apparent that the "PORTING NOTES" applied.